### PR TITLE
Respect shell globs when running `docker cp`.

### LINKED ISSUE #6013: Respect Shell Globs when running Docker cp...or make a sensible error.
I was trying to copy some files out of a container with globs.
```
[ec2-user@ip-172-31-17-218 ~]$ sudo docker cp jenkins:/opt/jenkins/data/com* .
2014/05/23 16:28:40 Tarball too short
[ec2-user@ip-172-31-17-218 ~]$ sudo docker cp jenkins:/opt/jenkins/data/* .
2014/05/23 16:28:51 Tarball too short
[ec2-user@ip-172-31-17-218 ~]$ sudo docker cp jenkins:/opt/jenkins/data/ .
```
`Tarball too short` is sort of meaningless....I didn't know I was using tar. There were files matching the `com*` pattern. There were files and directories matching the `*` pattern.

### PR DESCRIPTION
**Pull request by calavera**
Fixes #6013

/cc @huslage, @jlhawn

Signed-off-by: David Calavera david.calavera@gmail.com

### CODE REVIEW

**Comment by duglin** [comment]
+1 to the idea.
Would it make sense to use go's glob stuff so we support more than just * - like we do in the builder code?

**Comment by calavera** [comment]
@duglin can you point me to the builder part you're referring too? I'm slightly lost. I don't know if you saw that the patch uses go's `filepath.Glob` for file expansion, I don't know if that's what you meant.

**Comment by duglin** [comment]
@calavera I was just skimming the PR and noticed that https://github.com/docker/docker/pull/12858/files#diff-6ce9e79ddb91a3f06352abe1f2c72ecbR990 only looked for "*" and I think it would need to look for more than that if it was to support the full wildcard set.

**Comment by duglin** [comment]
Haven't played with Glob() but I bet you could just use it even in the non-wildcard case and it'll still return a match - then you don't need to check to see if wildcards are used at all.

**Comment by calavera** [comment]
> I bet you could just use it even in the non-wildcard case and it'll still return a match

yeah, I thought about that last night, we can probably remove the condition and use glob to match anything. I'll change the code to reflect that.

**Comment by calavera** [comment]
@duglin I've unified this logic to always use glob and added test to verify that other matching expressions work as expected. The tests for the old behavior still pass :tada:

**Comment by duglin** [comment]
excellent! will review in a little bit....

File: daemon/container.go [hunk 1]
```diff
@@ -1557,3 +1548,37 @@ func (c *Container) LogDriverType() string {
...
+	if err != nil {
+		return nil, err
+	}
+
+	if len(all) == 0 {
+		return nil, os.ErrNotExist
+	}
+
+	baseDir := filepath.Dir(basePath)
+	for baseDir != "" && !strings.HasPrefix(all[0], baseDir) {
```
Comment (duglin) [inline daemon/container.go]: Can you explain when you would need to make the baseDir up one level?

File: daemon/container.go [hunk 2]
```diff
@@ -1557,3 +1548,37 @@ func (c *Container) LogDriverType() string {
...
+		baseDir = filepath.Dir(baseDir)
+	}
+
+	relative := make([]string, len(all))
+
+	p := fmt.Sprintf("%s%s", baseDir, string(filepath.Separator))
+	for i, f := range all {
+		r := strings.Replace(f, p, "", -1)
```
Comment (duglin) [inline daemon/container.go]: I think doing `r := f[len(p):]` would be faster and safer since the `-1` could force the replacement to happen multiple times in for the path, and not just the leading bit.
```
l := len(fmt.Sprintf("%s%s", baseDir,string(filepath.Separator))
for ...
   r := f[l:]
   ...
```

File: daemon/container.go (hunk 1 above)
Comment (calavera) [inline daemon/container.go]: Imagine that the matcher is `/tmp/**/*`. You want to find the base directory common to the expanded paths, for instance `/tmp/foo/test`, and the matcher. Otherwise, you could end up using `/tmp/**` as a base directory, and that doesn't exist.

File: integration-cli/docker_cli_cp_test.go [hunk 3]
```diff
@@ -596,3 +596,66 @@ func (s *DockerSuite) TestCpNameHasColon(c *check.C) {
...
+
+func (s *DockerSuite) TestCpGlob(c *check.C) {
+	out, exitCode := dockerCmd(c, "run", "-d", "busybox", "/bin/sh", "-c", "mkdir -p '"+cpTestPath+"' && echo -n '"+cpContainerContents+"' > "+cpFullPath)
+	if exitCode != 0 {
+		c.Fatal("failed to create a container", out)
+	}
+
+	cleanedContainerID := strings.TrimSpace(out)
+
+	out, _ = dockerCmd(c, "wait", cleanedContainerID)
```
Comment (duglin) [inline integration-cli/docker_cli_cp_test.go]: I think it would be easier to just remove the `-d` from the `docker run` and name the container, then use that name instead of the id later on.

File: integration-cli/docker_cli_cp_test.go [hunk 4]
```diff
@@ -596,3 +596,66 @@ func (s *DockerSuite) TestCpNameHasColon(c *check.C) {
...
+	if strings.TrimSpace(out) != "0" {
+		c.Fatal("failed to set up container", out)
+	}
+
+	if err := os.MkdirAll(cpTestPath, os.ModeDir); err != nil {
```
Comment (duglin) [inline integration-cli/docker_cli_cp_test.go]: I can't figure out what the "hostFile" is used for - can you explain?
You appear to open it, write to it but don't close it until the defer is run.  But I also can't seem to see you actually using that file later on either.

File: integration-cli/docker_cli_cp_test.go [hunk 5]
```diff
@@ -596,3 +596,66 @@ func (s *DockerSuite) TestCpNameHasColon(c *check.C) {
...
+			c.Errorf("output matched host file -- %s - %s\n", cs.pathMatch, cs.tmpName)
+		}
+
+		if string(test) != cpContainerContents {
+			c.Errorf("output doesn't match the input -- %s - %s\n", cs.pathMatch, cs.tmpName)
+		}
+
+		os.RemoveAll(filepath.Join(tmpdir, "*"))
+	}
+}
```
Comment (duglin) [inline integration-cli/docker_cli_cp_test.go]: Can we get a test for the case where it glob doesn't match any file?

File: pkg/symlink/fs_test.go [hunk 6]
```diff
@@ -400,3 +400,17 @@ func TestFollowSymlinkNoLexicalCleaning(t *testing.T) {
 		t.Fatal(err)
 	}
 }
+
+func TestFollowSymlinkWildcard(t *testing.T) {
```
Comment (duglin) [inline pkg/symlink/fs_test.go]: Is this new test related to this PR?

File: daemon/container.go (hunk 2 above)
Comment (calavera) [inline daemon/container.go]: aggreed

File: integration-cli/docker_cli_cp_test.go (hunk 3 above)
Comment (calavera) [inline integration-cli/docker_cli_cp_test.go]: I'm just copying how we initialize other containers for integration tests. I rather not doing anything different.

File: integration-cli/docker_cli_cp_test.go (hunk 4 above)
Comment (calavera) [inline integration-cli/docker_cli_cp_test.go]: It might not be used, I copied the initialization from other cp tests. I'll double check.

File: integration-cli/docker_cli_cp_test.go (hunk 5 above)
Comment (calavera) [inline integration-cli/docker_cli_cp_test.go]: totally :+1:

File: pkg/symlink/fs_test.go (hunk 6 above)
Comment (calavera) [inline pkg/symlink/fs_test.go]: it is. Just to make sure that the symlink library doesn't try to expand glob matches and put them in right context.

File: daemon/container.go (hunk 1 above)
Comment (duglin) [inline daemon/container.go]: hmm if my container has:
```
/a/b/c/d/e/file
```
and I ask for /a/b/c/d/e/file  I get back just 'file'
but if I ask for /a/b/*/*/*/file I get back "c/d/e/file".  
I'm not sure but is this really what should happen?  Should we get different results when using "*" even if it matches just a single path?

Just brainstorming, but I wonder if perhaps instead it should look for the longest common path across all of the files and use that as the root dir.  For example, the result list was:
```
/a/b/c/d/e/file
/a/b/file
```
this would result in a common path of "/a/b" and therefore the paths returns to the CLI would be:
```
c/d/e/file
file
```

Comment (calavera) [inline daemon/container.go]: yeah, I think you're right. The problem here is that there are not enough tests to cover other behaviors with more than one file. I'll change the logic to reflect that.

**Comment by calavera** [comment]
@duglin I've changed the implementation to match the path behavior you described, it makes way more sense. I've also added tests for that behavior and not found paths.

Let me know what you think.

File: daemon/container.go [hunk 7]
```diff
@@ -1627,3 +1618,53 @@ func (container *Container) monitorExec(execConfig *execConfig, callback execdri
...
+
+func findSignificantParent(paths []string) string {
+	baseDir := filepath.Dir(paths[0])
+
+	for i := 1; i < len(paths); i++ {
+		p := paths[i]
+		if strings.HasPrefix(baseDir, p) {
+			baseDir = p
+		} else if !strings.HasPrefix(p, baseDir) {
+			d1 := filepath.Dir(baseDir)
```
Comment (duglin) [inline daemon/container.go]: Would you need to keep moving up the tree looking for common paths instead of just one level?  I would think that if baseDir was "/a/b/c" and p was "/d/e/f" then baseDir after this if-statement would need to be just "/".  Right?

File: daemon/container_unit_test.go [hunk 8]
```diff
@@ -195,3 +196,31 @@ func TestValidContainerNames(t *testing.T) {
...
+	}
+
+	paths = []string{
+		"/etc/foo",
+		"/etc/bar/baz",
+	}
+	if b := findSignificantParent(paths); b != "/etc" {
+		t.Fatalf("Expected /etc, was %s\n", b)
+	}
+}
```
Comment (duglin) [inline daemon/container_unit_test.go]: Let's add a test for things like:
```
/etc/foo/bar
/tmp/boogie/nights
```
to make sure it comes up with `/`.

And:
```
/tmp/dir1/dir2/dir3/dir4/file
/tmp/dir1/dir3/dir4/dir5/file
```
results in `/tmp/dir1` to make sure we walk up more than one level when needed.

File: daemon/container.go [hunk 9]
```diff
@@ -1627,3 +1618,53 @@ func (container *Container) monitorExec(execConfig *execConfig, callback execdri
...
+
+	if len(paths) == 0 {
+		return nil, os.ErrNotExist
+	}
+
+	baseDir := findSignificantParent(paths)
+
+	relative := make([]string, len(paths))
+
+	p := fmt.Sprintf("%s%s", baseDir, string(filepath.Separator))
```
Comment (duglin) [inline daemon/container.go]: I think you need an if-statement around the adding of "/" so that if baseDir is just "/" you don't add another one.

**Comment by duglin** [comment]
I haven't been able to figure out what's going on but I do see something kind of neat/funky  :-)
```
$curl -s -H "Content-Type:application/json" -X POST -d '{"Resource":"/*"}' http://127.0.0.1:2376/v1.19/containers/bee37b85c512/copy  | tar -tf- | head -10
.dockerenv
.dockerinit
a/
a/b/
a/b/c/
a/b/c/d/
a/b/c/d/e/
a/b/c/d/e/file
...

But then look at this:
$curl -s -H "Content-Type:application/json" -X POST -d '{"Resource":"/"}' http://127.0.0.1:2376/v1.19/containers/bee37b85c512/copy  | tar -tf- | head -10
bee37b85c512f864f46b01c12bf33e3e80cbb286a3b91ef69051c03b788d27a8/
bee37b85c512f864f46b01c12bf33e3e80cbb286a3b91ef69051c03b788d27a8/.dockerenv
bee37b85c512f864f46b01c12bf33e3e80cbb286a3b91ef69051c03b788d27a8/.dockerinit
bee37b85c512f864f46b01c12bf33e3e80cbb286a3b91ef69051c03b788d27a8/a/
bee37b85c512f864f46b01c12bf33e3e80cbb286a3b91ef69051c03b788d27a8/a/b/
bee37b85c512f864f46b01c12bf33e3e80cbb286a3b91ef69051c03b788d27a8/a/b/c/
bee37b85c512f864f46b01c12bf33e3e80cbb286a3b91ef69051c03b788d27a8/a/b/c/d/
...
```
I don't know for sure but I'm guessing this might have something to do with this line:
https://github.com/docker/docker/pull/12858/files?diff=unified#diff-6ce9e79ddb91a3f06352abe1f2c72ecbR1645
and then passing in "" as an IncludeFile into the archiver.  But didn't dig any deeper yet....

File: daemon/container.go [hunk 10]
```diff
@@ -1627,3 +1618,53 @@ func (container *Container) monitorExec(execConfig *execConfig, callback execdri
...
+	for i, f := range paths {
+		r := f[len(p):]
+
+		if r != "" {
```
Comment (duglin) [inline daemon/container.go]: When would we expect r to be "" ?

**Comment by calavera** [comment]
@duglin I decided to go with another implementation(a pseudo-tree) to find the right path and it looks like it works properly. I've added more test cases that should cover the scenarios you describe.

File: daemon/container.go [hunk 11]
```diff
@@ -1627,3 +1618,93 @@ func (container *Container) monitorExec(execConfig *execConfig, callback execdri
...
+	relative := make([]string, len(paths))
+
+	for i, f := range paths {
+		relative[i] = f[len(baseDir):]
+	}
+
+	return &archiveFilter{baseDir, relative}, nil
+}
+
+func findSignificantParent(paths []string) string {
```
Comment (duglin) [inline daemon/container.go]: This seems like a lot of work.  Would something a little smaller like this work?  https://gist.github.com/duglin/c88c3d5fde9c9ac6fd22

**Comment by calavera** [comment]
@duglin it looks like everything works as expected with your patch :tada:

File: integration-cli/docker_cli_cp_test.go [hunk 12]
```diff
@@ -596,3 +596,82 @@ func (s *DockerSuite) TestCpNameHasColon(c *check.C) {
...
+	}
+
+	cleanedContainerID := strings.TrimSpace(out)
+
+	out, _ = dockerCmd(c, "wait", cleanedContainerID)
+	if strings.TrimSpace(out) != "0" {
+		c.Fatal("failed to set up container", out)
+	}
+
+	if err := os.MkdirAll(cpTestPath, os.ModeDir); err != nil {
```
Comment (duglin) [inline integration-cli/docker_cli_cp_test.go]: Is the stuff on lines 615 thru 619 needed?  I'm not sure why we would need to create these dirs on the local machine if we cp into "tmpdir" ?

File: integration-cli/docker_cli_cp_test.go [hunk 13]
```diff
@@ -596,3 +596,82 @@ func (s *DockerSuite) TestCpNameHasColon(c *check.C) {
...
+
+			if string(test) == cpHostContents {
+				c.Errorf("output matched host file -- %s - %s\n", cs.pathMatch, cs.tmpName)
+			}
+
+			if string(test) != cpContainerContents {
+				c.Errorf("output doesn't match the input -- %s - %s\n", cs.pathMatch, cs.tmpName)
+			}
+		}
+		os.RemoveAll(filepath.Join(tmpdir, "*"))
```
Comment (duglin) [inline integration-cli/docker_cli_cp_test.go]: Is this needed if we have the `defer os.RemoveAll(tmpdir)` above?

Comment (calavera) [inline integration-cli/docker_cli_cp_test.go]: yes, it cleans the directory between test cases.

**Comment by calavera** [comment]
I've cleaned up that dead code in the test.

File: integration-cli/docker_cli_cp_test.go (hunk 13 above)
Comment (duglin) [inline integration-cli/docker_cli_cp_test.go]: ok

Comment (duglin) [inline integration-cli/docker_cli_cp_test.go]: Are you sure "*" is ok in the path? I didn't notice the docs saying it deal with globs.  You might be better off just erasing all of tmpdir and then recreating it before each iteration.

Comment (duglin) [inline integration-cli/docker_cli_cp_test.go]: I meant "*" in the RemoveAll arg.

**Comment by calavera** [comment]
@duglin it looks like your algorithm didn't handle copying `/` correctly either, but it was caused for a previous cleanup that we do in the input. I'm generating temp directories for every test case now and it looks like it works as expected.

File: daemon/container.go [hunk 14]
```diff
@@ -1034,28 +1034,23 @@ func (container *Container) Copy(resource string) (io.ReadCloser, error) {
 		}
 	}()
 
+	if len(resource) == 0 { // Copying `/` we expand to copy its content
```
Comment (duglin) [inline daemon/container.go]: Not sure, but are you sure we wouldn't be better off generating an error? Its like a "cp" w/o a src path - kind of odd. And some people may think that "" should mean "current dir" and not "/".

Comment (duglin) [inline daemon/container.go]: If we do end up keeping this then let's add a test for it in TestCpGlob()

**Comment by duglin** [comment]
just one comment - otherwise its looking good

File: daemon/container.go (hunk 14 above)
Comment (calavera) [inline daemon/container.go]: this actually is the case when you do `docker cp ID:/ /tmp`, copying the whole file system. We strip the first `/` at the api level and arrive to this point with an empty string. Docker already fails if you try to copy an empty string, but I've added a test case to be explicit about it.

Comment (duglin) [inline daemon/container.go]: hmm didn't realize the api layer did that - I kind of wonder why.  Oh well.

**Comment by duglin** [comment]
LGTM

**Comment by jlhawn** [comment]
I haven't reviewed this yet, but I expect a major conflict will occur with https://github.com/docker/docker/pull/13171 :sob:

**Comment by calavera** [comment]
@jlhawn conflicts are always solvable. This only cares about the source, regardless you're copying to the host or other container.

**Comment by calavera** [comment]
@jlhawn does your PR implements this feature then? Let me know to close this, I don't want to have it around if it's duplicated.

**Comment by duglin** [comment]
Right now it doesn't but that would be really cool if it did - hint hint :-)

**Comment by LK4D4** [comment]
@calavera would you mind to rebase?

**Comment by calavera** [comment]
@LK4D4 rebased

**Comment by icecrime** [comment]
LGTM

File: docs/reference/commandline/cli.md [hunk 15]
```diff
@@ -976,6 +976,9 @@ relative to the root of the container's filesystem.
 
     Copy files/folders from the PATH to the HOSTDIR.
 
+This command respects shell globs inside the container's filesystem. For instance:
+
+    docker cp CONTAINER:/tmp/* /tmp
```
Comment (thaJeztah) [inline docs/reference/commandline/cli.md]: I think wildcards may be a more widely used term than "shell globs". Perhaps;

```Markdown
The `docker cp` command supports the use of wildcards to specify which
files and folders to copy from the container's filesystem.

For example, run this command to copy all files and folders in the
`/tmp/` directory inside the container to the `/home/foo/` directory
on the host:

    $ docker cp CONTAINER:/tmp/* /home/foo/
```

**Comment by thaJeztah** [comment]
Thanks @calavera! I added a comment inline. Can you also update the [man/docker-cp.1.md](https://github.com/docker/docker/blob/master/man/docker-cp.1.md) file to include this information?

ping @fredlf @jamtur01 @moxiegirl @SvenDowideit

**Comment by duglin** [comment]
I'm a bit confused. If we merged this, and then merged #13171 won't we regress in function?  @jlhawn are you planning on adding glob support soon?

**Comment by jlhawn** [comment]
I would prefer to add it separately, as it would make #13171 even larger than it already is.

Also, I'm not 100% on board with the current design of this feature. I'm not a fan of overloading the `copy`/`archive-path` endpoints with the ability to handle globs, perform a search for the most significant parent, and return an archive of *that* common parent directory with all of the relative paths included. This means that the resulting archive will have entries with arbitrary paths depending on the common parent and it doesn't work with what I have currently going on in #13171.

I have suggested another approach where we add a separate endpoint that returns paths in the container which matches a shell glob. The `cp` command in the client can this to expand the source argument. Then all we need to do is implement `docker cp` with multiple sources.

**Comment by jlhawn** [comment]
@duglin @icecrime @calavera Should I go ahead and add what I described to #13171?

**Comment by calavera** [comment]
I'm closing this PR. I don't want to be biased on my review of #13171.

**Comment by duglin** [comment]
I'm ok with not doing the "most significant parent" thing (and basing the results on / instead), but I'm not sure I understand the value of having a separate endpoint just to expand wildcards and then forcing the client to have to issue a second command. 

By doing that you're asking the client to inspect the input, look for all possible wildcards (and deal with escaping), and then do extra processing based on the outcome of that analysis.  What makes it more tricky is that if we decide to add support for a different set of wildcards then each and every client would need to change to match it.  

Unless, you're expecting all callers to first call the new "expand globs API" and then use that result in the "copy files" API? In which case, if calling "expand" is (pretty much) required, why not just always do it under "copy files" for people?  If people really have usecases for not expanding wildcards then we can add a flag to disable it.  Just doesn't seem a user friendly to split it to me.

**Comment by jlhawn** [comment]
> Unless, you're expecting all callers to first call the new "expand globs API" and then use that result in the "copy files" API? In which case, if calling "expand" is (pretty much) required, why not just always do it under "copy files" for people?

Yes, I was thinking of *always* having the server expand wildcards. I don't think the API should be a 'copy' API, but it should provide the needed combination of endpoints to make it possible. The added round trips for the multiple API calls is negligible, in my opinion.

> If people really have usecases for not expanding wildcards then we can add a flag to disable it.

Or have users of the `docker cp` command properly escape any glob characters that they don't want the server to expand. This is also what `scp` requires.

> Just doesn't seem a user friendly to split it to me.

User friendly for API users or for CLI users? I think that API users would appreciate the simplicity and understand the separation of functionality. And CLI users would appreciate the familiarity of `docker cp` to `scp`.

**Comment by jlhawn** [comment]
I also want to point out that my original PR for `docker cp` (back in January) worked the other way: it actually **was** a copy API, but several of the design reviewers didn't like that approach because it required the server to have too much extra context about the source (when copying to a container) or destination (when copying from a container). I agreed with the arguments that were made and those discussions resulted in the current design with a much simpler archive/extract API where the client would be required to prepare the archives for the desired copy behavior.

**Comment by KES777** [comment]
can not copy from container:

```
$ sudo docker cp maitre_d_postgres:'/opt/monkeyman/bin/ur_2/*5*' ./
Error response from daemon: lstat /var/lib/docker/overlay2/668f3659d6e5212c815170067d8b66ab553355b26ee551273139ecdf387f10f3/merged/opt/monkeyman/bin/ur_2/*5*: no such file or directory
```
